### DISCUSSION TITLE
More data for custom label printing mixin

### DISCUSSION THREAD

**Discussion by luwol03** [inventree/InvenTree#3368] [Q&A] [answered]
Hey, 

what would be the best way to get more data about a part/location/stock item via a custom `LabelPrintingMixin`? My first thought was to use some external library to decode the QR to text and the use the inventree API to fetch more info about the part which can then be used to be also printed on the label. But this feels like a bad workaround. Can you please let me know if there is no other option and how do I make sure, that the QR code library which is needed for decoding will be installed on plugin install?

**Comment by matmair** [comment, marked as answer]
@luwol03 I am not sure why you would need extra information in a plugin exposing LabelPrintingMixin APIs. This is generally meant to only take the pdf / picture and send it to *some kind of* printer. Kind of like a printer driver in an OS.
Maybe it will help my understanding if you give a small example what you want to achive. If you want to add information to a label my suggestion would be to change the label itself, you have access to nearly everything in the templating engine.
Plugins are generally meant to be distributed as a python packages (either via PyPi, a VCS or a private index like githubs) , you just need to add the required libraries as a requirement in the setup (see the deployment advice in the [docs](https://inventree.readthedocs.io/en/latest/extend/plugins/#plugin-installation-file-pip) / the [packaging page](https://inventree.readthedocs.io/en/latest/extend/how_to_plugin/#packaging))

**Reply by luwol03** [reply]
What do you mean by this? Where do I find that?
> If you want to add information to a label my suggestion would be to change the label itself, you have access to nearly everything in the templating engine.

### Usecase
Im trying to print labels like this to attach them to little ESD boxes with my components to also manually see what's inside. My printer needs this in a custom XML format (Dymo - you can read more about that [here](https://www.matteomattei.com/use-python-to-print-labels-with-dymo-labelwriter-450/) if you want), so its not possible to use a complete picture scaled over the whole label.
![image](https://user-images.githubusercontent.com/60048565/180067645-bb5ffa0b-4904-4210-86a4-4de4ba1051b2.png)

**Reply by matmair** [reply]
@luwol03 intresting. Looks a bit like Zebras ZPL.
I am not sure if the LabelPrintingMixin ist the right interface for handeling this. I think it might be wise to implement separate functions for handing over the raw information so that this can be handled properly. @SchrodingersGat thoughts / a prefered way?

**Reply by SchrodingersGat** [reply]
@luwol03 an interesting problem. The current label printing framework only allows for printing of pre-rendered labels, and does not assume that the plugin has an independent way of rendering label data.

This means that the label templates are *not* going to be rendering into a format that your printer supports...

I would suggest that the best approach here is to write a completely separate plugin which provides the custom functionality you want, and generates labels for your printer. You could either do this as a plugin directly integrated into the server, or using the (external) python API to retrieve the data.

**Reply by luwol03** [reply]
Ok. No, I wanted to use the existing printing buttons. Then I try my workaround and come back if I have any questions.

How does your labes look like? Do they only contain the QR code?

**Reply by SchrodingersGat** [reply]
> How does your labes look like? Do they only contain the QR code? 

Labels can contain qrcodes, barcodes, text, images, whatever you want really. The "out of the box" labels have qrcodes and text. But users certainly can customize and adapt them however they want.

Are you sure that the dymo printer cannot just receive a PDF? https://www.dymo-label-printers.co.uk/news/faq-can-i-print-pdf-files-on-the-labelwriter.html

**Reply by luwol03** [reply]
Not sure, have to test. But if he can, where can I adjust the templates?

**Reply by SchrodingersGat** [reply]
https://inventree.readthedocs.io/en/latest/report/report/

**Reply by luwol03** [reply]
1. @SchrodingersGat What about giving the ability to parse arbitrary rendered label templates to the `LabelPrintingMixin` instead of limit it to a PNG file? Then these templates could also return XML or any data which can then be used by the `LabelPrintingMixin` to be sent to the printer (e.g. also XML files).

2. While searching where this is implemented, I found that there are actually handed over more arguments than I can access. Do I miss some part? I'd appreciate it, if you could show me the relevant code lines.
From what I got, the plugin function is called by this [line](https://github.com/inventree/InvenTree/blob/master/InvenTree/plugin/base/label/label.py#L47-L55) with much more kwargs, also described [here](https://github.com/inventree/InvenTree/blob/0c97a50e47728466089c0463d5518bf267177de6/InvenTree/plugin/base/label/mixins.py#L26-L33), but looking at the output I get, there is something strange.
```py
class MyLabelPrinter(LabelPrintingMixin, InvenTreePlugin):
    NAME = "MyLabelPrinter"
    SLUG = "mylabel"
    TITLE = "A dummy printer"

    def print_label(self, label, *args ,**kwargs):
        print("-> PRINT <-", label, args, kwargs)
```
```bash
inventree-inventree-worker-1  | -> PRINT <- <PIL.PngImagePlugin.PngImageFile image mode=RGBA size=2481x3508 at 0x7F363AC809D0> <class 'PIL.PngImagePlugin.PngImageFile'> () {'width': 24.0, 'height': 24.0}
```

**Reply by matmair** [reply]
@luwol03 iam not totally sure what the question in 2) is but you can  access all the data by replacing the catchall variables at the end ( *args, *kwargs) with the names in the function call you linked.

As for 1): the Api was designed to handle the printing of one or more rendered labels. Maybe it is cleaner to add a new interface for this to keep the number of variables small.

**Reply by luwol03** [reply]
1. I thought about getting an array of rendered templates in whatever format (e.g. XML documents or pdf files, ...) for each label one entry which can than be handled by the PrintigMixin how it's splittend and sent to the printer. E.g. a pdf file will just be made into a Grid and printed by a A4 printer, or the renderer XML gets printed one by one by a label printer, ... why is that template limited to a png image instead of letting the plugin writer decide based on the template what he wants.

2. @matmair In the linked code lines, there are specified much more kwargs as the ones printed to the console. So I think there are some parts in the code that I haven't found. I'd asked if you can list all relevant code parts which gets called after each other if I click on print in the web ui.

**Reply by SchrodingersGat** [reply]
The label printer plugins can use either the .pdf or .PNG versions of the rendered label. 

The backend rendering pipeline uses weasyprint which takes templated HTML and generates a pdf file. There is no way to generate XML from this, and even if so it negates all of the layout information from our templates. The whole idea behind rendering to pdf is to be portable and mostly universal.

Your printer has a specific XML format, there's no guarantee other printers would follow that. If you can get it accept a pdf file please use that.

I don't see a good way forward to supporting specific printer formats in inventree

**Reply by luwol03** [reply]
@SchrodingersGat My problem Stil is, how can I access all of these parameters I linked a few posts ago. If I log kwargs, I only get width and height. And label is typof a PIL image. Where is the pdf file?

**Reply by SchrodingersGat** [reply]
If you are running off of `stable` branch your items might be limited. Label printing has been significantly updated for 0.8.0.

https://github.com/inventree/InvenTree/blob/master/InvenTree/plugin/base/label/label.py

**Reply by luwol03** [reply]
Oh, this might be the problem. Thank you in advance. I'll try that out later. With pdf I could build a better workaround if I want.

**Reply by luwol03** [reply]
I now run latest and tried to access some new kwargs. Especially the `label_instance` in my own `LabelPrintingMixin` like shown below:
```py
print(kwargs['label_instance']) # Part Label - Simple part label
print(type(kwargs['label_instance'])) # <class 'label.models.PartLabel'>
``` 

then I tried to use the `get_context_data` method as it is described [here](https://github.com/inventree/InvenTree/blob/d32054b53b79f4603b2b06d4adc1e834557e55a4/InvenTree/label/models.py#L308) on the `PartLabel` class. But if I try to call that, I get the following errors and nothing works anymore.

```py
print(kwargs['label_instance'].get_context_data())
```

### Error

#### UI
![image](https://user-images.githubusercontent.com/60048565/180621532-f50dbe58-8be6-4c6a-a6e6-ec58d72fbffe.png)

#### Log:
```bash
inventree-inventree-proxy-1   | 192.168.48.1 - - [23/Jul/2022:20:20:19 +0000] "GET /api/label/part/1/print/?parts[]=1&plugin=samplelabel1 HTTP/1.1" 200 51 "http://localhost:8000/part/1/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.5 Safari/605.1.15" "-"
inventree-inventree-worker-1  | 20:20:19 [Q] INFO Process-1:1 processing [monkey-jersey-violet-enemy]
inventree-inventree-worker-1  | Plugin 'samplelabel1' is printing a label 'label.pdf'
inventree-inventree-worker-1  | Saving setting 'InvenTreeSetting:LABEL_DPI' to cache
inventree-inventree-worker-1  | Printing Label 123: label.pdf (User: admin)
inventree-inventree-worker-1  | Saving setting 'InvenTreeSetting:INVENTREE_BASE_URL' to cache
inventree-inventree-worker-1  | /root/.local/lib/python3.9/site-packages/django/db/models/fields/__init__.py:1358: RuntimeWarning: DateTimeField NotificationEntry.updated received a naive datetime (2022-07-22 00:00:00) while time zone support is active.
inventree-inventree-worker-1  |   warnings.warn("DateTimeField %s.%s received a naive datetime "
inventree-inventree-worker-1  | Gathering users for notification 'inventree.error_log'
inventree-inventree-worker-1  | Sending notification 'inventree.error_log' for 'Error object (39)'
inventree-inventree-worker-1  | Triggering notification method 'ui_message'
inventree-inventree-worker-1  | Notify users via 'ui_message' for notification 'inventree.error_log' for 'Error object (39)'
inventree-inventree-worker-1  | Event triggered: 'common_notificationmessage.created'
inventree-inventree-worker-1  | 20:20:19 [Q] INFO Enqueued 1019
inventree-inventree-worker-1  | Notified 1 users via 'ui_message' for notification 'inventree.error_log' for 'Error object (39)' successfully
inventree-inventree-worker-1  | Label printing failed: Sending notification to user 'admin'
inventree-inventree-worker-1  | Notification 'label.printing_failed' has recently been sent for 'Label Printer 1 - samplelabel1' - SKIPPING
inventree-inventree-proxy-1   | 192.168.48.1 - - [23/Jul/2022:20:20:19 +0000] "GET /api/notifications/?read=false HTTP/1.1" 200 1021 "http://localhost:8000/part/1/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.5 Safari/605.1.15" "-"
inventree-inventree-worker-1  | 20:20:19 [Q] INFO Processed [monkey-jersey-violet-enemy]
inventree-inventree-worker-1  | 20:20:19 [Q] INFO Process-1:3 processing [robert-yellow-freddie-cola]
inventree-inventree-worker-1  | Registering triggered event: 'common_notificationmessage.created'
inventree-inventree-worker-1  | Saving setting 'InvenTreeSetting:ENABLE_PLUGINS_EVENTS' to cache
inventree-inventree-worker-1  | 20:20:19 [Q] INFO Processed [robert-yellow-freddie-cola]
inventree-inventree-proxy-1   | 192.168.48.1 - - [23/Jul/2022:20:20:24 +0000] "GET /api/notifications/?read=false HTTP/1.1" 200 1032 "http://localhost:8000/part/1/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.5 Safari/605.1.15" "-"
```


### Version

<details>

```
# Version Information:
InvenTree-Version: 0.8.0 dev
Django Version: 3.2.14
Commit Hash: d32054b
Commit Date: 2022-07-22
Database: postgresql
Debug-Mode: True
Deployed using Docker: True
```

</details>

**Reply by SchrodingersGat** [reply]
@luwol03 to retrieve the full error message you will need to navigate to the "admin" page and share the error report - https://inventree.readthedocs.io/en/latest/settings/logs/

**Reply by luwol03** [reply]
@SchrodingersGat, sorry that I forgot this. This is the error I got.
<details>
<summary>Error</summary>

```py
Traceback (most recent call last):

File "/home/inventree/InvenTree/plugin/base/label/label.py", line 47, in print_label
plugin.print_label(

File "/home/inventree/InvenTree/plugins/print_label.py", line 24, in print_label
print("-> LABEL - INSTANCE DATA 123", kwargs["label_instance"].get_context_data())

TypeError: get_context_data() missing 1 required positional argument: 'request'
```

</details>

Then I noticed, that I need to hand over `None` manually for the `request` argument, which is ok I think, because [here](https://github.com/inventree/InvenTree/blob/d32054b53b79f4603b2b06d4adc1e834557e55a4/InvenTree/InvenTree/helpers.py#L323) its also set to `None` if not defined. After doing this, this is the new error:

<details>
<summary>Error</summary>

```py
Traceback (most recent call last):

File "/home/inventree/InvenTree/plugin/base/label/label.py", line 47, in print_label
plugin.print_label(

File "/home/inventree/InvenTree/plugins/print_label.py", line 24, in print_label
print("-> LABEL - INSTANCE DATA 123", kwargs["label_instance"].get_context_data(None))

File "/home/inventree/InvenTree/label/models.py", line 314, in get_context_data
'category': part.category,

AttributeError: 'NoneType' object has no attribute 'category'
``` 

</details>

but then I get this error. It seems like there is no label instance handed over with the `self.object_to_print` attribute. So I tried to analyze the call stack:

<details>
<summary>Call stack</summary>

From what I got, the call stak is as follows:

1. https://github.com/inventree/InvenTree/blob/cf70e4220f4c032f473e3282c45b650277d2187e/InvenTree/label/api.py#L500 gets called by the webUI via a GET request
2. because this class extends the `PartLabelMixin`, the following `get_parts()` method get called https://github.com/inventree/InvenTree/blob/cf70e4220f4c032f473e3282c45b650277d2187e/InvenTree/label/api.py#L418 which fetches all parts from the db provided in the queryparams.
3. because the class extends the `LabelPrintMixin` with https://github.com/inventree/InvenTree/blob/cf70e4220f4c032f473e3282c45b650277d2187e/InvenTree/label/api.py#L504 this method gets called: https://github.com/inventree/InvenTree/blob/cf70e4220f4c032f473e3282c45b650277d2187e/InvenTree/label/api.py#L80
4. then all outputs are constructed with its own **correct** label instance https://github.com/inventree/InvenTree/blob/cf70e4220f4c032f473e3282c45b650277d2187e/InvenTree/label/api.py#L99-L111
5. then **one** new dummy label_instance is constructed (it has no `object_to_print` https://github.com/inventree/InvenTree/blob/cf70e4220f4c032f473e3282c45b650277d2187e/InvenTree/label/api.py#L127
6. and then you iterate over all output items and always provide the same label_instance with the different rendered pdf files to the `offload_task` method https://github.com/inventree/InvenTree/blob/cf70e4220f4c032f473e3282c45b650277d2187e/InvenTree/label/api.py#L129-L143
7. the offloaded task then calls https://github.com/inventree/InvenTree/blob/d32054b53b79f4603b2b06d4adc1e834557e55a4/InvenTree/plugin/base/label/label.py#L18
8. which then calls my plugin https://github.com/inventree/InvenTree/blob/d32054b53b79f4603b2b06d4adc1e834557e55a4/InvenTree/plugin/base/label/label.py#L47-L55

</details>

### Patch

So I suggest the following patch for step 5. If you like that, should I create an PR for that?

<details>

```patch
diff --git a/InvenTree/label/api.py b/InvenTree/label/api.py
index 7b533c882..68c420517 100644
--- a/InvenTree/label/api.py
+++ b/InvenTree/label/api.py
@@ -95,6 +95,7 @@ class LabelPrintMixin:
         label_name = "label.pdf"
 
         label_names = []
+        label_instances = []
 
         # Merge one or more PDF files into a single download
         for item in items_to_print:
@@ -104,6 +105,7 @@ class LabelPrintMixin:
             label_name = label.generate_filename(request)
 
             label_names.append(label_name)
+            label_instances.append(label)
 
             if debug_mode:
                 outputs.append(label.render_as_string(request))
@@ -123,9 +125,6 @@ class LabelPrintMixin:
             - Return a JSON response indicating that the printing has been offloaded
             """
 
-            # Label instance
-            label_instance = self.get_object()
-
             for idx, output in enumerate(outputs):
                 """For each output, we generate a temporary image file, which will then get sent to the printer."""
 
@@ -138,7 +137,7 @@ class LabelPrintMixin:
                     plugin.plugin_slug(),
                     pdf,
                     filename=label_names[idx],
-                    label_instance=label_instance,
+                    label_instance=label_instances[idx],
                     user=request.user,
                 )
```

</details>

**Reply by SchrodingersGat** [reply]
@luwol03 seems like a reasonable patch, allowing you to fully recreate the label context in a plugin, if required.

**Reply by matmair** [reply]
Patch looks good

**Reply by luwol03** [reply]
Patched in #3349 for reference.
